site-src/geps/gep-851.md

```diff
@@ -0,0 +1,88 @@
+# GEP-851: Allow Multiple Certificate Refs per Gateway Listener
+
+* Issue: [#851](https://github.com/kubernetes-sigs/gateway-api/issues/851)
+* Status: Implementable
+
+## TLDR
+
+Replace `CertificateRef` field with an `CertificateRefs` field in Gateway
+Listeners.
+
+## Goals
+
+* Allow certs to be attached for multiple unique Hostnames for the same
+  Listener. This is most helpful when the hostname is empty on the Listener and
+  attached Routes specify multiple hostnames.
+* Enable specifying RSA and ECDSA certs on the same Listener.
+* Improve cert upgrades by allowing new and old certs to be attached to the same
+  Listener.
+
+## Non-Goals
+
+* Be too prescriptive about how multiple certificates are handled for a Gateway
+  Listener. Each implementation may handle this slightly differently.
+
+## Introduction
+
+When we removed certificate references from HTTPRoute ([GEP
+746](/geps/gep-746/)) we lost the capability to attach multiple certificates to
+a Listener. This GEP proposes adding that capability back, directly to the
+Gateway Listener this time.
+
+In addition to helping with the self service approach, where app owners may
+attach multiple hostnames (and certs) to a single Listener, this can help with
+other use cases:
+
+* Attaching RSA and ECDSA certs.
+* Attaching both new and old certs during an upgrade.
+
+## API
+
+The `CertificateRef` field in `GatewayTLSConfig` would be replaced with the
+following `CertificateRefs` field:
+
+```go
+	// CertificateRefs contains a series of references to Kubernetes objects that
+	// contains TLS certificates and private keys. These certificates are used to
+	// establish a TLS handshake for requests that match the hostname of the
+	// associated listener.
+  //
+  // When multiple TLS certificates are specified, SNI matching MUST be used to
```

Comment:
So, the behavior here is that you can specify as many certs as you like, and they'll only be used if one of the cert's CN or SAN fields contains a match for a hostname in either the Listener or the Route?

Are implementations expected to check the CN and SAN of all certs and tell users if something doesn't match up? If they do, should the Gateway API present a standard way to indicate that in status/Conditions? I kind of feel like the best user experience will be if implementations do have to check, and tell you if there's zero valid matches between TLS, Listener Hostname, and Route hostname.

Reply:
> Are implementations expected to check the CN and SAN of all certs and tell users if something doesn't match up? If they do, should the Gateway API present a standard way to indicate that in status/Conditions

I think it makes sense, we should check the CN and SAN with the hostname in either the Listener or the Route.

Reply:
> Are implementations expected to check the CN and SAN of all certs and tell users if something doesn't match up? If they do, should the Gateway API present a standard way to indicate that in status/Conditions? I kind of feel like the best user experience will be if implementations do have to check, and tell you if there's zero valid matches between TLS, Listener Hostname, and Route hostname.

+1 with one exception:

Sometimes, users configure a self-signed certificate (with no or some random CN/SANs) for all of their listeners, really common with dev environments and some other times when TLS trust is not important. Another case is when you initially start with cert-manager as well I think (at least that used to be a thing). How do we handle for those cases?

To handle the above, we currently use listener hostname -> TLS cert to establish the relationship between the SNI and a TLS cert (and don't use TLS SNI in the handshake to look up the cert).

Reply:
It seems like we all agree that we should have a way to communicate/warn when the CN and SAN do not overlap with any values that are allowed by the listener and attached routes. Based on @hbagdi's scenario above, it seems like we need to leave some flexibility as far as what implementations do when that happens. I think it's reasonable to require that implementations add a warning when these conflict, but I can't think of a good way to require implementations to actually prevent this while still allowing for Harry's use case.

Reply:
I updated this to match what I said above, let me know what you think @hbagdi @tokers @youngnick.